### PR TITLE
Fixed #28718 -- Allowed password reset if a user's current password doesn't use an enabled hasher.

### PR DESCRIPTION
**Pull request by timgraham**
Regression in aeb1389442d0f9669edf6660b747fd10693b63a7.
Reverted changes to is_password_usable() from
703c266682be39f7153498ad0d8031231f12ee79 and documentation changes from
92f48680dbd2e02f2b33f6ad0e35b7d337889fb2.

### CODE REVIEW

File: django/contrib/auth/hashers.py [hunk 1]
```diff
@@ -21,13 +21,8 @@
...
 def is_password_usable(encoded):
-    if encoded is None or encoded.startswith(UNUSABLE_PASSWORD_PREFIX):
-        return False
-    try:
-        identify_hasher(encoded)
-    except ValueError:
-        return False
-    return True
+    """Is the password something besides make_password(None)?"""
+    return encoded is None or not encoded.startswith(UNUSABLE_PASSWORD_PREFIX)
```
Comment (apollo13) [inline django/contrib/auth/hashers.py]: `None` is supposed to be a usable password?

Comment (claudep) [reply django/contrib/auth/hashers.py]: I'm not fond either of this semantic change of the `is_password_usable()` API. I think we need two APIs for two different use cases: 1. tell if the current user password is usable (as current `is_usable_password`), 2. tell if the current user password is voluntarily disabled.

Comment (timgraham) [reply django/contrib/auth/hashers.py]: Looking at 703c266682be39f7153498ad0d8031231f12ee79 again, I see that `None` was previously treated as a disabled password. There were no tests or documentation about the rationale for that however. 

Claude, https://github.com/django/django/pull/9276 proposed a separate method as you suggested, however, I think that would leave `is_usable_password()` unused by Django as the two places it's used, I believe we want the semantics of "was `set_unusable_password()` called?" I don't see a use case for the current semantics, besides in `check_password()` which is adapted here. I updated the documentation for `User.has_usable_password()` only a couple months ago in (92f48680dbd2e02f2b33f6ad0e35b7d337889fb2 when I was looking at this issue previously and I thought a separate API was needed, but I don't think your change in 703c266682be39f7153498ad0d8031231f12ee79 was entirely intentional as you didn't update the documentation at that time. Feel free to elaborate on your thinking.

Comment (claudep) [reply django/contrib/auth/hashers.py]: Fair, but then I think a `versionchanged` is missing in the docs + maybe a backwards incompatibility release note.

Comment (timgraham) [reply django/contrib/auth/hashers.py]: Sure, I've added that.

File: django/contrib/auth/hashers.py [hunk 2]
```diff
@@ -42,7 +40,11 @@ def check_password(password, encoded, setter=None, preferred='default'):
         return False
 
     preferred = get_hasher(preferred)
-    hasher = identify_hasher(encoded)
+    try:
+        hasher = identify_hasher(encoded)
+    except ValueError:
+        # encoded uses a password hasher that's no longer installed.
```
[the remark this answers is not in the document]
Comment (timgraham) [reply django/contrib/auth/hashers.py]: I updated the comment. I don't think it's worth additional complexity above.

File: docs/releases/2.1.txt [hunk 3]
```diff
@@ -358,6 +358,15 @@ Miscellaneous
   changed from 0 to an empty string, which mainly may require some adjustments
   in tests that compare HTML.
 
+* :meth:`.User.has_usable_password` and the
+  :func:`~django.contrib.auth.hashers.is_password_usable` function no longer
+  return ``False`` if the password is ``None``, or if the password uses a
```
[the remark this answers is not in the document]
Comment (timgraham) [reply docs/releases/2.1.txt]: Sure

**Comment by MarkusH** [comment]
What problem are we trying to solve? As I understand the ticket, active users with a password from a password hasher that's no longer known, are not able to reset the password. Right? Can we just allow that and still treat `None`, `''` and `!garbish` as unusable passwords?

**Comment by timgraham** [comment]
Yes, but I don't see a compelling reason to. If `None` and `''` mean unusable password, then why do we have the "starts with exclamation mark" case? It could be useful, for example, to allow creating a user with a null password and then ask the user to request a password reset. If we treat `None` as unusable, then this case won't work.

**Comment by MarkusH** [comment]
> It could be useful, for example, to allow creating a user with a null password and then ask the user to request a password reset. If we treat `None` as unusable, then this case won't work.

Good use case. Thanks

File: tests/auth_tests/test_hashers.py [hunk 4]
```diff
@@ -277,8 +277,9 @@ def test_bad_algorithm(self):
             identify_hasher('lolcat$salt$hash')
 
     def test_bad_encoded(self):
```
Comment (akulakov) [inline tests/auth_tests/test_hashers.py]: Should this now be `test_password_usable()` ?

**Pull request merged**